File: JavaScript/EncodeAndDecodeTinyURL/EncodeAndDecodeTinyURL.js
/**
 * Encodes a URL to a shortened URL.
 *
 * @param {string} longUrl
 * @return {string}
 */
// (152ms)
var hashCode = function(str){
    var hash = 0;
    if (str.length == 0) return hash;
    for (i = 0; i < str.length; i++) {
        char = str.charCodeAt(i);
        hash = ((hash<<5)-hash)+char;
        hash = hash & hash; // Convert to 32bit integer
    }
    return hash;
}

var cache = {}

var encode = function(longUrl) {
    var key = hashCode(longUrl).toString;
    cache[key] = longUrl;
    return key;
};

/**
 * Decodes a shortened URL to its original URL.
 *
 * @param {string} shortUrl
 * @return {string}
 */
var decode = function(shortUrl) {
    return cache[shortUrl];
};

/**
 * Your functions will be called as such:
 * decode(encode(url));
 */

File: Python/EncodeAndDecodeTinyURL/EncodeAndDecodeTinyURL.py
class Codec:

    # (49ms)
    cache = {}

    def encode(self, longUrl):
        """Encodes a URL to a shortened URL.

        :type longUrl: str
        :rtype: str
        """
        key = hash(longUrl)
        self.cache[key] = longUrl
        return str(key)


    def decode(self, shortUrl):
        """Decodes a shortened URL to its original URL.

        :type shortUrl: str
        :rtype: str
        """
        key = int(shortUrl)
        return self.cache[key]



        # Your Codec object will be instantiated and called as such:
        # codec = Codec()
        # codec.decode(codec.encode(url))